aip/0151.md

```diff
@@ -0,0 +1,86 @@
+---
+aip:
+  id: 151
+  state: reviewing
+  created: 2019-07-25
+permalink: /151
+redirect_from:
+  - /0151
+---
+
+# Long-running operations
+
+Occasionally, an API may need to expose a method that takes a significant
+amount of time to complete. In these situations, it is often a poor user
+experience to simply block while the task runs; rather, it is better to return
+some kind of promise to the user and allow the user to check back in later.
+
+The long-running operations pattern is roughly analogous to a [Python
+`Future`][], or a [Node.js `Promise`][]. Essentially, the user is given a token
+that can be used to track progress and retrieve the result.
+
+## Guidance
+
+Individual API methods that might take a significant amount of time to complete
+**should** return a [`google.longrunning.Operation`][lro] object instead of the
+ultimate response message.
+
+```proto
+// Write a book.
+rpc WriteBook(WriteBookRequest) returns (google.longrunning.Operation) {
+  option (google.api.http) = {
+    post: "/v1/{parent=shelves/*}/books}:write"
+    body: "*"
+  };
+  option (google.longrunning.operation_info) = {
+    response_type: "WriteBookResponse"
+    metadata_type: "WriteBookMetadata"
+  };
+}
+```
+
+- The response type **must** be `google.longrunning.Operation`. The `Operation`
+  proto definition **must not** be copied into individual APIs.
+- The method **must** include a `google.longrunning.operation_info` annotation
+  defining the response and metadata types.
+  - The response and metadata types **must** be defined in the file where the
+    RPC appears, or a file imported by that file.
+  - The metadata type **should not** be `google.protobuf.Empty`, as this
+    prevents metadata from being added later. Instead, define an empty message
+    for the RPC and use that.
+- APIs with messages that return `Operation` **must** implement the
+  [`Operations`][lro] service. Individual APIs **must not** define their own
+  interfaces for long-running operations to avoid inconsistency.
+
+**Note:** User expectations can vary on what is considered "a significant
+amount of time" depending on what work is being done. A good rule of thumb is
+10 seconds.
+
+### Standard methods
+
+APIs **may** return an `Operation` from the [`Create`][aip-133],
+[`Update`][aip-134], or [`Delete`][aip-135] standard methods if appropriate. In
+this case, the response type in the `operation_info` annotation **must** be the
+standard and expected response type for that standard method.
+
+When creating or deleting a resource with a long-running operation, the
+resource **should** be included in List and Get calls; however, the resource
```

Comment:
List and Get with code font, and possibly link to AIPs as paragraph above?

Reply:
Done.